lib/src/highlighter.dart

```diff
@@ -44,28 +47,496 @@ class Highlighter {
+  /// file that aren't adjacent in the original file.
+  static bool _contiguous(List<_Line> lines) {
+    for (var i = 0; i < lines.length - 1; i++) {
+      var thisLine = lines[i];
+      var nextLine = lines[i + 1];
+      if (thisLine.number + 1 != nextLine.number &&
+          thisLine.url == nextLine.url) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /// Collect all the source lines from the contexts of all spans in
+  /// [highlights], and associates them with the highlights that cover them.
+  static List<_Line> _collateLines(List<_Highlight> highlights) {
+    _sortHighlights(highlights);
+
+    var lines = <_Line>[];
+
+    // First, populate [lines] with all the lines that we have context for along
+    // with their line numbers.
+    Uri lastUrl;
+    for (var highlight in highlights) {
+      var context = highlight.span.context;
+      // If [highlight.span.context] contains lines prior to the one
+      // [highlight.span.text] appears on, write those first.
+      var lineStart = findLineStart(
+          context, highlight.span.text, highlight.span.start.column);
+      assert(lineStart != null); // enforced by [_normalizeContext]
+
+      var linesBeforeSpan =
+          '\n'.allMatches(context.substring(0, lineStart)).length;
+
+      var url = highlight.span.sourceUrl;
+      var lineNumber = highlight.span.start.line - linesBeforeSpan;
+      for (var line in context.split('\n')) {
+        // Only add a line if it hasn't already been added for a previous span.
+        if (lines.isEmpty || lineNumber > lines.last.number || url != lastUrl) {
+          lines.add(_Line(line, lineNumber, url));
+        }
+        lineNumber++;
+      }
+
+      lastUrl = url;
+    }
+
+    // Next, associate each line with each highlights that covers it.
+    var activeHighlights = <_Highlight>[];
+    var highlightIndex = 0;
+    for (var line in lines) {
+      activeHighlights.removeWhere((highlight) =>
+          highlight.span.sourceUrl != line.url ||
+          highlight.span.end.line < line.number);
+
+      var oldHighlightLength = activeHighlights.length;
+      for (var highlight in highlights.skip(highlightIndex)) {
+        if (highlight.span.start.line > line.number) break;
+        if (highlight.span.sourceUrl != line.url) break;
+        activeHighlights.add(highlight);
+      }
+      highlightIndex += activeHighlights.length - oldHighlightLength;
+
+      line.highlights.addAll(activeHighlights);
+    }
+
+    return lines;
+  }
+
+  /// Sorts [highlights] in-place so that highlights of the same files are
+  /// contiguous, the primary highlight's file appears first, and all highlights
+  /// within a given file are ordered according to their spans.
+  static void _sortHighlights(List<_Highlight> highlights) {
+    if (highlights.length == 1) return;
+
+    // Track the existing order of files to disrupt it as little as possible.
```

Comment:
Some of the code here feels to me like it would be easier to read if we keep a `Map<Uri, List<_Highlight>>` instead of a `List<_Highlight>`. All the `url != lastUrl` and related patterns would not be necessary.

A `LinkedHashMap` will preserve key iteration order and give us most of the rest of this for free. I think we also know that the primary highlight is always the first in the list?


```dart
var partitionedHighlights = {highlights.first.span.sourceUrl: [highlights.first]};
for(var highlight in highlights.skip(1)) {
  (partitionedHighlights[highlight.span.sourcesUrl] ??= []).add(highlight);
}
for(var partition in partitionedHighlights.values) {
  partition.sort();
}
```

Reply:
Good idea, done.